src/Result.ts

```diff
@@ -1,115 +1,150 @@
+/** Object with information about the issue identified by the validation rules  */
+export interface Issue {
+	/** The message with information about this issue */
+	message: string;
+}
+
 /** Top-level result, returned from a validator function */
-export interface Result {
+export class Result {
 	/** Equivalent of the previous return type, the full collection of error messages (including errors from child properties) */
-	errorMessages: string[];
+	get errorMessages(): string[] {
```

Comment:
@JoshuaKGoldberg  This addresses some feedback you gave in the original design conversation about maintaining two forms of state for the same thing.  That issue would have become more a problem with this class-based approach (e.g. ChildResult.addIssue after it's already been added to a parent Result wouldn't update the parent's `errorMessages`).  So, I went ahead and took care of it through an accessor that derives it on the fly.  That way we maintain feature parity and expose the convenience prop, while not maintaining it in two places.